Reactor hooks optimizations

robotmrv opened the pull request:
Share results of my work on Sleuth Reactor instrumentation optimization  

This is WIP. There are several places which use `reactor-core` internals, but I hope they could be resolved (one of probable solution is described here https://github.com/reactor/reactor-core/issues/2242#issuecomment-683346691).  

Mostly it optimizes `DECORATE_ON_EACH` instrumentation by using `reactor-core` feature from https://github.com/reactor/reactor-core/issues/2058 (added `Scannable.Attr.RunStyle`) (Which was the follow up of #1553)  
cc @adriancole @simonbasle (as they participate in original issue)  

Result of benchmark https://github.com/spring-cloud/spring-cloud-sleuth/blob/master/benchmarks/src/test/java/org/springframework/cloud/sleuth/benchmarks/jmh/webflux/MicroBenchmarkHttpTests.java to compare effect:
current master as baseline: 
```
Benchmark                                                       (instrumentation)    Mode    Cnt      Score        Error   Units
MicroBenchmarkHttpTests.test                                       noSleuthSimple  sample  86801      0.115 ±      0.001   ms/op
MicroBenchmarkHttpTests.test:test·p0.00                            noSleuthSimple  sample             0.055                ms/op
MicroBenchmarkHttpTests.test:test·p0.50                            noSleuthSimple  sample             0.123                ms/op
MicroBenchmarkHttpTests.test:test·p0.90                            noSleuthSimple  sample             0.131                ms/op
MicroBenchmarkHttpTests.test:test·p0.95                            noSleuthSimple  sample             0.137                ms/op
MicroBenchmarkHttpTests.test:test·p0.99                            noSleuthSimple  sample             0.163                ms/op
MicroBenchmarkHttpTests.test:test·p0.999                           noSleuthSimple  sample             0.213                ms/op
MicroBenchmarkHttpTests.test:test·p0.9999                          noSleuthSimple  sample             0.391                ms/op
MicroBenchmarkHttpTests.test:test·p1.00                            noSleuthSimple  sample             3.731                ms/op
MicroBenchmarkHttpTests.test:·gc.alloc.rate                        noSleuthSimple  sample     10    168.406 ±     69.358  MB/sec
MicroBenchmarkHttpTests.test:·gc.alloc.rate.norm                   noSleuthSimple  sample     10  34528.748 ±     74.930    B/op
MicroBenchmarkHttpTests.test:·gc.churn.PS_Eden_Space               noSleuthSimple  sample     10    154.906 ±    184.454  MB/sec
MicroBenchmarkHttpTests.test:·gc.churn.PS_Eden_Space.norm          noSleuthSimple  sample     10  33895.297 ±  35449.919    B/op
MicroBenchmarkHttpTests.test:·gc.count                             noSleuthSimple  sample     10      7.000               counts
MicroBenchmarkHttpTests.test:·gc.time                              noSleuthSimple  sample     10     23.000                   ms
MicroBenchmarkHttpTests.test                                   sleuthSimpleManual  sample  70213      0.142 ±      0.001   ms/op
MicroBenchmarkHttpTests.test:test·p0.00                        sleuthSimpleManual  sample             0.076                ms/op
MicroBenchmarkHttpTests.test:test·p0.50                        sleuthSimpleManual  sample             0.148                ms/op
MicroBenchmarkHttpTests.test:test·p0.90                        sleuthSimpleManual  sample             0.158                ms/op
MicroBenchmarkHttpTests.test:test·p0.95                        sleuthSimpleManual  sample             0.172                ms/op
MicroBenchmarkHttpTests.test:test·p0.99                        sleuthSimpleManual  sample             0.198                ms/op
MicroBenchmarkHttpTests.test:test·p0.999                       sleuthSimpleManual  sample             0.257                ms/op
MicroBenchmarkHttpTests.test:test·p0.9999                      sleuthSimpleManual  sample             0.375                ms/op
MicroBenchmarkHttpTests.test:test·p1.00                        sleuthSimpleManual  sample             5.587                ms/op
MicroBenchmarkHttpTests.test:·gc.alloc.rate                    sleuthSimpleManual  sample     10    190.331 ±      3.873  MB/sec
MicroBenchmarkHttpTests.test:·gc.alloc.rate.norm               sleuthSimpleManual  sample     10  42760.732 ±    131.501    B/op
MicroBenchmarkHttpTests.test:·gc.churn.PS_Eden_Space           sleuthSimpleManual  sample     10    184.031 ±    240.895  MB/sec
MicroBenchmarkHttpTests.test:·gc.churn.PS_Eden_Space.norm      sleuthSimpleManual  sample     10  41587.725 ±  54432.434    B/op
MicroBenchmarkHttpTests.test:·gc.count                         sleuthSimpleManual  sample     10      6.000               counts
MicroBenchmarkHttpTests.test:·gc.time                          sleuthSimpleManual  sample     10     29.000                   ms
MicroBenchmarkHttpTests.test                                         sleuthManual  sample  69978      0.143 ±      0.001   ms/op
MicroBenchmarkHttpTests.test:test·p0.00                              sleuthManual  sample             0.096                ms/op
MicroBenchmarkHttpTests.test:test·p0.50                              sleuthManual  sample             0.148                ms/op
MicroBenchmarkHttpTests.test:test·p0.90                              sleuthManual  sample             0.159                ms/op
MicroBenchmarkHttpTests.test:test·p0.95                              sleuthManual  sample             0.174                ms/op
MicroBenchmarkHttpTests.test:test·p0.99                              sleuthManual  sample             0.200                ms/op
MicroBenchmarkHttpTests.test:test·p0.999                             sleuthManual  sample             0.259                ms/op
MicroBenchmarkHttpTests.test:test·p0.9999                            sleuthManual  sample             0.375                ms/op
MicroBenchmarkHttpTests.test:test·p1.00                              sleuthManual  sample             5.726                ms/op
MicroBenchmarkHttpTests.test:·gc.alloc.rate                          sleuthManual  sample     10    191.422 ±      3.551  MB/sec
MicroBenchmarkHttpTests.test:·gc.alloc.rate.norm                     sleuthManual  sample     10  43147.201 ±    111.042    B/op
MicroBenchmarkHttpTests.test:·gc.churn.PS_Eden_Space                 sleuthManual  sample     10    180.962 ±    236.601  MB/sec
MicroBenchmarkHttpTests.test:·gc.churn.PS_Eden_Space.norm            sleuthManual  sample     10  40988.366 ±  53604.625    B/op
MicroBenchmarkHttpTests.test:·gc.churn.PS_Survivor_Space             sleuthManual  sample     10      0.002 ±      0.007  MB/sec
MicroBenchmarkHttpTests.test:·gc.churn.PS_Survivor_Space.norm        sleuthManual  sample     10      0.466 ±      1.484    B/op
MicroBenchmarkHttpTests.test:·gc.count                               sleuthManual  sample     10      6.000               counts
MicroBenchmarkHttpTests.test:·gc.time                                sleuthManual  sample     10     30.000                   ms
MicroBenchmarkHttpTests.test                                   sleuthSimpleOnEach  sample  52990      0.189 ±      0.001   ms/op
MicroBenchmarkHttpTests.test:test·p0.00                        sleuthSimpleOnEach  sample             0.134                ms/op
MicroBenchmarkHttpTests.test:test·p0.50                        sleuthSimpleOnEach  sample             0.191                ms/op
MicroBenchmarkHttpTests.test:test·p0.90                        sleuthSimpleOnEach  sample             0.211                ms/op
MicroBenchmarkHttpTests.test:test·p0.95                        sleuthSimpleOnEach  sample             0.224                ms/op
MicroBenchmarkHttpTests.test:test·p0.99                        sleuthSimpleOnEach  sample             0.257                ms/op
MicroBenchmarkHttpTests.test:test·p0.999                       sleuthSimpleOnEach  sample             0.321                ms/op
MicroBenchmarkHttpTests.test:test·p0.9999                      sleuthSimpleOnEach  sample             5.214                ms/op
MicroBenchmarkHttpTests.test:test·p1.00                        sleuthSimpleOnEach  sample             7.406                ms/op
MicroBenchmarkHttpTests.test:·gc.alloc.rate                    sleuthSimpleOnEach  sample     10    181.365 ±      3.175  MB/sec
MicroBenchmarkHttpTests.test:·gc.alloc.rate.norm               sleuthSimpleOnEach  sample     10  53991.521 ±    178.747    B/op
MicroBenchmarkHttpTests.test:·gc.churn.PS_Eden_Space           sleuthSimpleOnEach  sample     10    213.213 ±    225.669  MB/sec
MicroBenchmarkHttpTests.test:·gc.churn.PS_Eden_Space.norm      sleuthSimpleOnEach  sample     10  63907.694 ±  67760.995    B/op
MicroBenchmarkHttpTests.test:·gc.count                         sleuthSimpleOnEach  sample     10      7.000               counts
MicroBenchmarkHttpTests.test:·gc.time                          sleuthSimpleOnEach  sample     10     36.000                   ms
MicroBenchmarkHttpTests.test                                   sleuthSimpleOnLast  sample  64342      0.155 ±      0.001   ms/op
MicroBenchmarkHttpTests.test:test·p0.00                        sleuthSimpleOnLast  sample             0.097                ms/op
MicroBenchmarkHttpTests.test:test·p0.50                        sleuthSimpleOnLast  sample             0.159                ms/op
MicroBenchmarkHttpTests.test:test·p0.90                        sleuthSimpleOnLast  sample             0.180                ms/op
MicroBenchmarkHttpTests.test:test·p0.95                        sleuthSimpleOnLast  sample             0.188                ms/op
MicroBenchmarkHttpTests.test:test·p0.99                        sleuthSimpleOnLast  sample             0.218                ms/op
MicroBenchmarkHttpTests.test:test·p0.999                       sleuthSimpleOnLast  sample             0.279                ms/op
MicroBenchmarkHttpTests.test:test·p0.9999                      sleuthSimpleOnLast  sample             1.638                ms/op
MicroBenchmarkHttpTests.test:test·p1.00                        sleuthSimpleOnLast  sample             5.104                ms/op
MicroBenchmarkHttpTests.test:·gc.alloc.rate                    sleuthSimpleOnLast  sample     10    186.558 ±     11.177  MB/sec
MicroBenchmarkHttpTests.test:·gc.alloc.rate.norm               sleuthSimpleOnLast  sample     10  45741.341 ±    148.094    B/op
MicroBenchmarkHttpTests.test:·gc.churn.PS_Eden_Space           sleuthSimpleOnLast  sample     10    146.143 ±    303.286  MB/sec
MicroBenchmarkHttpTests.test:·gc.churn.PS_Eden_Space.norm      sleuthSimpleOnLast  sample     10  35825.783 ±  74235.927    B/op
MicroBenchmarkHttpTests.test:·gc.churn.PS_Survivor_Space       sleuthSimpleOnLast  sample     10      0.028 ±      0.136  MB/sec
MicroBenchmarkHttpTests.test:·gc.churn.PS_Survivor_Space.norm  sleuthSimpleOnLast  sample     10      6.922 ±     33.093    B/op
MicroBenchmarkHttpTests.test:·gc.count                         sleuthSimpleOnLast  sample     10      4.000               counts
MicroBenchmarkHttpTests.test:·gc.time                          sleuthSimpleOnLast  sample     10     16.000                   ms
MicroBenchmarkHttpTests.test                                      noSleuthComplex  sample   4317      2.317 ±      0.023   ms/op
MicroBenchmarkHttpTests.test:test·p0.00                           noSleuthComplex  sample             1.241                ms/op
MicroBenchmarkHttpTests.test:test·p0.50                           noSleuthComplex  sample             2.038                ms/op
MicroBenchmarkHttpTests.test:test·p0.90                           noSleuthComplex  sample             3.035                ms/op
MicroBenchmarkHttpTests.test:test·p0.95                           noSleuthComplex  sample             3.117                ms/op
MicroBenchmarkHttpTests.test:test·p0.99                           noSleuthComplex  sample             3.150                ms/op
MicroBenchmarkHttpTests.test:test·p0.999                          noSleuthComplex  sample             3.221                ms/op
MicroBenchmarkHttpTests.test:test·p0.9999                         noSleuthComplex  sample             7.078                ms/op
MicroBenchmarkHttpTests.test:test·p1.00                           noSleuthComplex  sample             7.078                ms/op
MicroBenchmarkHttpTests.test:·gc.alloc.rate                       noSleuthComplex  sample     10      9.052 ±      3.569  MB/sec
MicroBenchmarkHttpTests.test:·gc.alloc.rate.norm                  noSleuthComplex  sample     10  37530.452 ±   1490.344    B/op
MicroBenchmarkHttpTests.test:·gc.churn.PS_Eden_Space              noSleuthComplex  sample     10     11.343 ±     36.153  MB/sec
MicroBenchmarkHttpTests.test:·gc.churn.PS_Eden_Space.norm         noSleuthComplex  sample     10  97208.953 ± 309854.228    B/op
MicroBenchmarkHttpTests.test:·gc.churn.PS_Survivor_Space          noSleuthComplex  sample     10      0.423 ±      1.350  MB/sec
MicroBenchmarkHttpTests.test:·gc.churn.PS_Survivor_Space.norm     noSleuthComplex  sample     10   3629.248 ±  11569.860    B/op
MicroBenchmarkHttpTests.test:·gc.count                            noSleuthComplex  sample     10      2.000               counts
MicroBenchmarkHttpTests.test:·gc.time                             noSleuthComplex  sample     10      7.000                   ms
MicroBenchmarkHttpTests.test                                        onEachComplex  sample   4305      2.326 ±      0.023   ms/op
MicroBenchmarkHttpTests.test:test·p0.00                             onEachComplex  sample             1.446                ms/op
MicroBenchmarkHttpTests.test:test·p0.50                             onEachComplex  sample             2.046                ms/op
MicroBenchmarkHttpTests.test:test·p0.90                             onEachComplex  sample             3.031                ms/op
MicroBenchmarkHttpTests.test:test·p0.95                             onEachComplex  sample             3.105                ms/op
MicroBenchmarkHttpTests.test:test·p0.99                             onEachComplex  sample             3.150                ms/op
MicroBenchmarkHttpTests.test:test·p0.999                            onEachComplex  sample             3.248                ms/op
MicroBenchmarkHttpTests.test:test·p0.9999                           onEachComplex  sample             3.457                ms/op
MicroBenchmarkHttpTests.test:test·p1.00                             onEachComplex  sample             3.457                ms/op
MicroBenchmarkHttpTests.test:·gc.alloc.rate                         onEachComplex  sample     10     15.780 ±      0.284  MB/sec
MicroBenchmarkHttpTests.test:·gc.alloc.rate.norm                    onEachComplex  sample     10  57900.485 ±   1859.159    B/op
MicroBenchmarkHttpTests.test:·gc.count                              onEachComplex  sample     10        ≈ 0               counts
MicroBenchmarkHttpTests.test                                        onLastComplex  sample   4304      2.324 ±      0.023   ms/op
MicroBenchmarkHttpTests.test:test·p0.00                             onLastComplex  sample             1.249                ms/op
MicroBenchmarkHttpTests.test:test·p0.50                             onLastComplex  sample             2.040                ms/op
MicroBenchmarkHttpTests.test:test·p0.90                             onLastComplex  sample             3.039                ms/op
MicroBenchmarkHttpTests.test:test·p0.95                             onLastComplex  sample             3.113                ms/op
MicroBenchmarkHttpTests.test:test·p0.99                             onLastComplex  sample             3.146                ms/op
MicroBenchmarkHttpTests.test:test·p0.999                            onLastComplex  sample             3.299                ms/op
MicroBenchmarkHttpTests.test:test·p0.9999                           onLastComplex  sample             3.719                ms/op
MicroBenchmarkHttpTests.test:test·p1.00                             onLastComplex  sample             3.719                ms/op
MicroBenchmarkHttpTests.test:·gc.alloc.rate                         onLastComplex  sample     10     13.347 ±      0.358  MB/sec
MicroBenchmarkHttpTests.test:·gc.alloc.rate.norm                    onLastComplex  sample     10  48940.329 ±   1868.315    B/op
MicroBenchmarkHttpTests.test:·gc.count                              onLastComplex  sample     10        ≈ 0               counts
MicroBenchmarkHttpTests.test                                      onManualComplex  sample   4296      2.328 ±      0.023   ms/op
MicroBenchmarkHttpTests.test:test·p0.00                           onManualComplex  sample             1.260                ms/op
MicroBenchmarkHttpTests.test:test·p0.50                           onManualComplex  sample             2.042                ms/op
MicroBenchmarkHttpTests.test:test·p0.90                           onManualComplex  sample             3.035                ms/op
MicroBenchmarkHttpTests.test:test·p0.95                           onManualComplex  sample             3.109                ms/op
MicroBenchmarkHttpTests.test:test·p0.99                           onManualComplex  sample             3.146                ms/op
MicroBenchmarkHttpTests.test:test·p0.999                          onManualComplex  sample             3.195                ms/op
MicroBenchmarkHttpTests.test:test·p0.9999                         onManualComplex  sample             3.547                ms/op
MicroBenchmarkHttpTests.test:test·p1.00                           onManualComplex  sample             3.547                ms/op
MicroBenchmarkHttpTests.test:·gc.alloc.rate                       onManualComplex  sample     10     12.731 ±      0.516  MB/sec
MicroBenchmarkHttpTests.test:·gc.alloc.rate.norm                  onManualComplex  sample     10  46755.414 ±   1768.817    B/op
MicroBenchmarkHttpTests.test:·gc.count                            onManualComplex  sample     10        ≈ 0               counts
```

results with changes:  
```
Benchmark                                                       (instrumentation)    Mode    Cnt      Score        Error   Units
MicroBenchmarkHttpTests.test                                       noSleuthSimple  sample  88950      0.112 ±      0.001   ms/op
MicroBenchmarkHttpTests.test:test·p0.00                            noSleuthSimple  sample             0.046                ms/op
MicroBenchmarkHttpTests.test:test·p0.50                            noSleuthSimple  sample             0.120                ms/op
MicroBenchmarkHttpTests.test:test·p0.90                            noSleuthSimple  sample             0.130                ms/op
MicroBenchmarkHttpTests.test:test·p0.95                            noSleuthSimple  sample             0.136                ms/op
MicroBenchmarkHttpTests.test:test·p0.99                            noSleuthSimple  sample             0.161                ms/op
MicroBenchmarkHttpTests.test:test·p0.999                           noSleuthSimple  sample             0.209                ms/op
MicroBenchmarkHttpTests.test:test·p0.9999                          noSleuthSimple  sample             0.442                ms/op
MicroBenchmarkHttpTests.test:test·p1.00                            noSleuthSimple  sample             3.994                ms/op
MicroBenchmarkHttpTests.test:·gc.alloc.rate                        noSleuthSimple  sample     10    172.642 ±     72.537  MB/sec
MicroBenchmarkHttpTests.test:·gc.alloc.rate.norm                   noSleuthSimple  sample     10  34487.748 ±     67.156    B/op
MicroBenchmarkHttpTests.test:·gc.churn.PS_Eden_Space               noSleuthSimple  sample     10    177.262 ±    201.426  MB/sec
MicroBenchmarkHttpTests.test:·gc.churn.PS_Eden_Space.norm          noSleuthSimple  sample     10  34087.974 ±  35923.281    B/op
MicroBenchmarkHttpTests.test:·gc.count                             noSleuthSimple  sample     10      7.000               counts
MicroBenchmarkHttpTests.test:·gc.time                              noSleuthSimple  sample     10     22.000                   ms
MicroBenchmarkHttpTests.test                                   sleuthSimpleManual  sample  70942      0.141 ±      0.001   ms/op
MicroBenchmarkHttpTests.test:test·p0.00                        sleuthSimpleManual  sample             0.077                ms/op
MicroBenchmarkHttpTests.test:test·p0.50                        sleuthSimpleManual  sample             0.147                ms/op
MicroBenchmarkHttpTests.test:test·p0.90                        sleuthSimpleManual  sample             0.156                ms/op
MicroBenchmarkHttpTests.test:test·p0.95                        sleuthSimpleManual  sample             0.170                ms/op
MicroBenchmarkHttpTests.test:test·p0.99                        sleuthSimpleManual  sample             0.196                ms/op
MicroBenchmarkHttpTests.test:test·p0.999                       sleuthSimpleManual  sample             0.255                ms/op
MicroBenchmarkHttpTests.test:test·p0.9999                      sleuthSimpleManual  sample             0.394                ms/op
MicroBenchmarkHttpTests.test:test·p1.00                        sleuthSimpleManual  sample             5.538                ms/op
MicroBenchmarkHttpTests.test:·gc.alloc.rate                    sleuthSimpleManual  sample     10    180.398 ±      2.511  MB/sec
MicroBenchmarkHttpTests.test:·gc.alloc.rate.norm               sleuthSimpleManual  sample     10  40122.316 ±    118.743    B/op
MicroBenchmarkHttpTests.test:·gc.churn.PS_Eden_Space           sleuthSimpleManual  sample     10    167.965 ±    219.985  MB/sec
MicroBenchmarkHttpTests.test:·gc.churn.PS_Eden_Space.norm      sleuthSimpleManual  sample     10  37448.797 ±  49089.375    B/op
MicroBenchmarkHttpTests.test:·gc.count                         sleuthSimpleManual  sample     10      6.000               counts
MicroBenchmarkHttpTests.test:·gc.time                          sleuthSimpleManual  sample     10     28.000                   ms
MicroBenchmarkHttpTests.test                                         sleuthManual  sample  69333      0.144 ±      0.001   ms/op
MicroBenchmarkHttpTests.test:test·p0.00                              sleuthManual  sample             0.085                ms/op
MicroBenchmarkHttpTests.test:test·p0.50                              sleuthManual  sample             0.149                ms/op
MicroBenchmarkHttpTests.test:test·p0.90                              sleuthManual  sample             0.165                ms/op
MicroBenchmarkHttpTests.test:test·p0.95                              sleuthManual  sample             0.174                ms/op
MicroBenchmarkHttpTests.test:test·p0.99                              sleuthManual  sample             0.200                ms/op
MicroBenchmarkHttpTests.test:test·p0.999                             sleuthManual  sample             0.252                ms/op
MicroBenchmarkHttpTests.test:test·p0.9999                            sleuthManual  sample             0.367                ms/op
MicroBenchmarkHttpTests.test:test·p1.00                              sleuthManual  sample             5.702                ms/op
MicroBenchmarkHttpTests.test:·gc.alloc.rate                          sleuthManual  sample     10    177.864 ±      8.906  MB/sec
MicroBenchmarkHttpTests.test:·gc.alloc.rate.norm                     sleuthManual  sample     10  40465.144 ±    113.512    B/op
MicroBenchmarkHttpTests.test:·gc.churn.PS_Eden_Space                 sleuthManual  sample     10    209.047 ±    369.025  MB/sec
MicroBenchmarkHttpTests.test:·gc.churn.PS_Eden_Space.norm            sleuthManual  sample     10  49124.407 ±  88471.096    B/op
MicroBenchmarkHttpTests.test:·gc.churn.PS_Survivor_Space             sleuthManual  sample     10      0.032 ±      0.145  MB/sec
MicroBenchmarkHttpTests.test:·gc.churn.PS_Survivor_Space.norm        sleuthManual  sample     10      7.457 ±     33.018    B/op
MicroBenchmarkHttpTests.test:·gc.count                               sleuthManual  sample     10      5.000               counts
MicroBenchmarkHttpTests.test:·gc.time                                sleuthManual  sample     10     19.000                   ms
MicroBenchmarkHttpTests.test                                   sleuthSimpleOnEach  sample  53947      0.185 ±      0.001   ms/op
MicroBenchmarkHttpTests.test:test·p0.00                        sleuthSimpleOnEach  sample             0.118                ms/op
MicroBenchmarkHttpTests.test:test·p0.50                        sleuthSimpleOnEach  sample             0.188                ms/op
MicroBenchmarkHttpTests.test:test·p0.90                        sleuthSimpleOnEach  sample             0.205                ms/op
MicroBenchmarkHttpTests.test:test·p0.95                        sleuthSimpleOnEach  sample             0.220                ms/op
MicroBenchmarkHttpTests.test:test·p0.99                        sleuthSimpleOnEach  sample             0.249                ms/op
MicroBenchmarkHttpTests.test:test·p0.999                       sleuthSimpleOnEach  sample             0.315                ms/op
MicroBenchmarkHttpTests.test:test·p0.9999                      sleuthSimpleOnEach  sample             0.434                ms/op
MicroBenchmarkHttpTests.test:test·p1.00                        sleuthSimpleOnEach  sample             5.931                ms/op
MicroBenchmarkHttpTests.test:·gc.alloc.rate                    sleuthSimpleOnEach  sample     10    120.930 ±     49.823  MB/sec
MicroBenchmarkHttpTests.test:·gc.alloc.rate.norm               sleuthSimpleOnEach  sample     10  39905.655 ±    134.834    B/op
MicroBenchmarkHttpTests.test:·gc.churn.PS_Eden_Space           sleuthSimpleOnEach  sample     10    125.304 ±    210.758  MB/sec
MicroBenchmarkHttpTests.test:·gc.churn.PS_Eden_Space.norm      sleuthSimpleOnEach  sample     10  41904.527 ±  66879.932    B/op
MicroBenchmarkHttpTests.test:·gc.count                         sleuthSimpleOnEach  sample     10      5.000               counts
MicroBenchmarkHttpTests.test:·gc.time                          sleuthSimpleOnEach  sample     10     25.000                   ms
MicroBenchmarkHttpTests.test                                   sleuthSimpleOnLast  sample  71197      0.140 ±      0.001   ms/op
MicroBenchmarkHttpTests.test:test·p0.00                        sleuthSimpleOnLast  sample             0.062                ms/op
MicroBenchmarkHttpTests.test:test·p0.50                        sleuthSimpleOnLast  sample             0.148                ms/op
MicroBenchmarkHttpTests.test:test·p0.90                        sleuthSimpleOnLast  sample             0.158                ms/op
MicroBenchmarkHttpTests.test:test·p0.95                        sleuthSimpleOnLast  sample             0.172                ms/op
MicroBenchmarkHttpTests.test:test·p0.99                        sleuthSimpleOnLast  sample             0.196                ms/op
MicroBenchmarkHttpTests.test:test·p0.999                       sleuthSimpleOnLast  sample             0.257                ms/op
MicroBenchmarkHttpTests.test:test·p0.9999                      sleuthSimpleOnLast  sample             4.470                ms/op
MicroBenchmarkHttpTests.test:test·p1.00                        sleuthSimpleOnLast  sample             5.964                ms/op
MicroBenchmarkHttpTests.test:·gc.alloc.rate                    sleuthSimpleOnLast  sample     10    193.766 ±     17.637  MB/sec
MicroBenchmarkHttpTests.test:·gc.alloc.rate.norm               sleuthSimpleOnLast  sample     10  42924.193 ±    159.213    B/op
MicroBenchmarkHttpTests.test:·gc.churn.PS_Eden_Space           sleuthSimpleOnLast  sample     10    202.098 ±    213.233  MB/sec
MicroBenchmarkHttpTests.test:·gc.churn.PS_Eden_Space.norm      sleuthSimpleOnLast  sample     10  46105.627 ±  48571.997    B/op
MicroBenchmarkHttpTests.test:·gc.count                         sleuthSimpleOnLast  sample     10      7.000               counts
MicroBenchmarkHttpTests.test:·gc.time                          sleuthSimpleOnLast  sample     10     35.000                   ms
MicroBenchmarkHttpTests.test                                      noSleuthComplex  sample   4313      2.319 ±      0.023   ms/op
MicroBenchmarkHttpTests.test:test·p0.00                           noSleuthComplex  sample             1.233                ms/op
MicroBenchmarkHttpTests.test:test·p0.50                           noSleuthComplex  sample             2.042                ms/op
MicroBenchmarkHttpTests.test:test·p0.90                           noSleuthComplex  sample             3.035                ms/op
MicroBenchmarkHttpTests.test:test·p0.95                           noSleuthComplex  sample             3.109                ms/op
MicroBenchmarkHttpTests.test:test·p0.99                           noSleuthComplex  sample             3.146                ms/op
MicroBenchmarkHttpTests.test:test·p0.999                          noSleuthComplex  sample             3.195                ms/op
MicroBenchmarkHttpTests.test:test·p0.9999                         noSleuthComplex  sample             5.603                ms/op
MicroBenchmarkHttpTests.test:test·p1.00                           noSleuthComplex  sample             5.603                ms/op
MicroBenchmarkHttpTests.test:·gc.alloc.rate                       noSleuthComplex  sample     10      9.031 ±      3.626  MB/sec
MicroBenchmarkHttpTests.test:·gc.alloc.rate.norm                  noSleuthComplex  sample     10  37395.502 ±    957.550    B/op
MicroBenchmarkHttpTests.test:·gc.churn.PS_Eden_Space              noSleuthComplex  sample     10     13.186 ±     63.043  MB/sec
MicroBenchmarkHttpTests.test:·gc.churn.PS_Eden_Space.norm         noSleuthComplex  sample     10  48508.890 ± 231916.761    B/op
MicroBenchmarkHttpTests.test:·gc.churn.PS_Survivor_Space          noSleuthComplex  sample     10      0.505 ±      2.412  MB/sec
MicroBenchmarkHttpTests.test:·gc.churn.PS_Survivor_Space.norm     noSleuthComplex  sample     10   1856.267 ±   8874.652    B/op
MicroBenchmarkHttpTests.test:·gc.count                            noSleuthComplex  sample     10      1.000               counts
MicroBenchmarkHttpTests.test:·gc.time                             noSleuthComplex  sample     10      3.000                   ms
MicroBenchmarkHttpTests.test                                        onEachComplex  sample   4298      2.327 ±      0.023   ms/op
MicroBenchmarkHttpTests.test:test·p0.00                             onEachComplex  sample             1.464                ms/op
MicroBenchmarkHttpTests.test:test·p0.50                             onEachComplex  sample             2.042                ms/op
MicroBenchmarkHttpTests.test:test·p0.90                             onEachComplex  sample             3.035                ms/op
MicroBenchmarkHttpTests.test:test·p0.95                             onEachComplex  sample             3.105                ms/op
MicroBenchmarkHttpTests.test:test·p0.99                             onEachComplex  sample             3.154                ms/op
MicroBenchmarkHttpTests.test:test·p0.999                            onEachComplex  sample             3.345                ms/op
MicroBenchmarkHttpTests.test:test·p0.9999                           onEachComplex  sample             3.518                ms/op
MicroBenchmarkHttpTests.test:test·p1.00                             onEachComplex  sample             3.518                ms/op
MicroBenchmarkHttpTests.test:·gc.alloc.rate                         onEachComplex  sample     10     10.400 ±      4.211  MB/sec
MicroBenchmarkHttpTests.test:·gc.alloc.rate.norm                    onEachComplex  sample     10  43173.387 ±   1111.996    B/op
MicroBenchmarkHttpTests.test:·gc.count                              onEachComplex  sample     10        ≈ 0               counts
MicroBenchmarkHttpTests.test                                        onLastComplex  sample   4329      2.310 ±      0.023   ms/op
MicroBenchmarkHttpTests.test:test·p0.00                             onLastComplex  sample             1.253                ms/op
MicroBenchmarkHttpTests.test:test·p0.50                             onLastComplex  sample             2.038                ms/op
MicroBenchmarkHttpTests.test:test·p0.90                             onLastComplex  sample             3.035                ms/op
MicroBenchmarkHttpTests.test:test·p0.95                             onLastComplex  sample             3.109                ms/op
MicroBenchmarkHttpTests.test:test·p0.99                             onLastComplex  sample             3.146                ms/op
MicroBenchmarkHttpTests.test:test·p0.999                            onLastComplex  sample             3.178                ms/op
MicroBenchmarkHttpTests.test:test·p0.9999                           onLastComplex  sample             3.199                ms/op
MicroBenchmarkHttpTests.test:test·p1.00                             onLastComplex  sample             3.199                ms/op
MicroBenchmarkHttpTests.test:·gc.alloc.rate                         onLastComplex  sample     10     12.622 ±      0.411  MB/sec
MicroBenchmarkHttpTests.test:·gc.alloc.rate.norm                    onLastComplex  sample     10  46008.979 ±   1882.009    B/op
MicroBenchmarkHttpTests.test:·gc.count                              onLastComplex  sample     10        ≈ 0               counts
MicroBenchmarkHttpTests.test                                      onManualComplex  sample   4310      2.321 ±      0.023   ms/op
MicroBenchmarkHttpTests.test:test·p0.00                           onManualComplex  sample             1.239                ms/op
MicroBenchmarkHttpTests.test:test·p0.50                           onManualComplex  sample             2.040                ms/op
MicroBenchmarkHttpTests.test:test·p0.90                           onManualComplex  sample             3.035                ms/op
MicroBenchmarkHttpTests.test:test·p0.95                           onManualComplex  sample             3.111                ms/op
MicroBenchmarkHttpTests.test:test·p0.99                           onManualComplex  sample             3.150                ms/op
MicroBenchmarkHttpTests.test:test·p0.999                          onManualComplex  sample             3.258                ms/op
MicroBenchmarkHttpTests.test:test·p0.9999                         onManualComplex  sample             3.387                ms/op
MicroBenchmarkHttpTests.test:test·p1.00                           onManualComplex  sample             3.387                ms/op
MicroBenchmarkHttpTests.test:·gc.alloc.rate                       onManualComplex  sample     10     11.960 ±      0.517  MB/sec
MicroBenchmarkHttpTests.test:·gc.alloc.rate.norm                  onManualComplex  sample     10  43792.191 ±   1854.119    B/op
MicroBenchmarkHttpTests.test:·gc.count                            onManualComplex  sample     10        ≈ 0               counts
```

Here some calculations:  
For `onEachComplex` case allocation reduced from plus `54.8%` to `15.45%` if compare with `noSleuthComplex`.  
`onEachComplex` allocates less than `onLastComplex` by `6.16%`, before changes `onEachComplex` allocated more by `25.8%`

adriancole commented on spring-cloud-sleuth-core/src/main/java/org/springframework/cloud/sleuth/instrument/reactor/Hack.java:
```diff
@@ -0,0 +1,450 @@
+	}
+
+	static boolean shouldDecorate(Publisher<?> p) {
+		int maxSteps = 100;
```
Please add a comment on this? What's the impact? What would make 10 or 1000 a bad number? are we trying to avoid recursion? or...

robotmrv replied:
The point of this is to protect from:  
- reduce overhead of traversal in long chains of `Publisher`s, but it seems would not work most of the time as there are no match reactive chains with such amount of operators. Probably such kind of check would helpful only on `Subscriber` level as their chains are much longer;  
- bad impls which could refers to wrong `source` so it could loop indefinitely;  

100 is just a magic number. The effect was not tested on long chains   
As I mentioned, this is WIP, it stayed here from previous implementation iteration. Probably should be deleted.

adriancole replied:
thanks for the notes.. if this persists, exactly what you wrote would be a fantastic comment!

adriancole commented on spring-cloud-sleuth-core/src/main/java/org/springframework/cloud/sleuth/instrument/reactor/Hack.java:
```diff
@@ -0,0 +1,450 @@
+ * 		Mono.just(it) // (-)
+ * 	) // (-)
+ * 	.flatMapMany(it ->
+ * 		asyncPublisher(it) // (+)(*)
```
This is an interesting one to explain as to the naked eye, it seems like it might drop the initial subscribe context. The other steps mentioned are more obvious and probably don't need more notes. You could add comments here or maybe a `<h3>Why asyncPublisher re-captures context from Brave</h3>` to free you from questions later.

robotmrv replied:
probably, to be more accurate this 
>(*) - captures brave context at subscription and propagates it  

could be rephrased to  
>(*) - captures brave context if it differs from what was captured before at subscription and propagates it. As far as check is performed on `Subscriber` `Context` it allocates `Publisher` to access it

adriancole commented on spring-cloud-sleuth-core/src/main/java/org/springframework/cloud/sleuth/instrument/reactor/Hack.java:
```diff
@@ -0,0 +1,450 @@
+
+}
+
+final class SleuthMonoLift<I, O> extends MonoOperator<I, O> implements SleuthDecorator {
```
As these types look like mechanical things to allow implementation of SleuthDecorator, it would be nice to add a one-line comment that "this does nothing special" or "this does XXX" to know the difference if any are doing something subtle.

robotmrv replied:
you are right, this is just a WA. Comment above is applicable to all this stuff   
> 	// WA need a way to determine whether Publisher is SleuthDecorator
> 	// mostly it is a copy of reactor-core logic
> 	// as on option - add Scannable to Lift Publishers which returns original lifter
> 	// function

I hope it would be deleted if `reactor-core` team adds access to the `lifter function`, so all this stuff would not be needed anymore.

adriancole commented on spring-cloud-sleuth-core/src/main/java/org/springframework/cloud/sleuth/instrument/reactor/ScopePassingSpanSubscriber.java:
```diff
@@ -115,6 +115,9 @@ public Object scanUnsafe(Attr key) {
 		if (key == Attr.PARENT) {
 			return this.s;
 		}
+		else if (key == Attr.RUN_STYLE) {
```
I wonder if there's an errorprone or other way to scan a project for a missing "if" like this (similar to missing branch in a case)...

adriancole commented on spring-cloud-sleuth-core/src/main/java/org/springframework/cloud/sleuth/instrument/reactor/SleuthDecorator.java:
```diff
@@ -0,0 +1,26 @@
+ * brave context into execution Thread.
+ */
+@Deprecated
+public interface SleuthDecorator {
```
At first I thought this was something else. Is it important to say Sleuth in the name? (wondering if the package is somehow missing in debug tools). If not, I would call it `TraceContextXXX` as this is similar to brave's rxjava2 propagation lib (except we didn't do operator fusion style dedupe there). The name `TraceContextXXX` is more precise as for example, this does not mutate spans or other features sleuth might do.

robotmrv replied:
`TraceContextPropagator` - is it fine?

adriancole commented on spring-cloud-sleuth-core/src/main/java/org/springframework/cloud/sleuth/instrument/web/TraceWebFilter.java:
```diff
@@ -111,6 +116,14 @@ Tracer tracer() {
 
+	CurrentTraceContext currentTraceContext() {
+		if (this.currentTraceContext == null) {
+			this.currentTraceContext = this.beanFactory.getBean(HttpTracing.class)
```
I think we only need the currentTraceContext, which has less strange wiring loop potential. Sometimes people use brave internally to make the Httptracing object. In some extreme cases, we guard the whole thing https://github.com/spring-cloud/spring-cloud-sleuth/blob/bd149ce4a7c2154fe8526394d14b6ab89addd62e/spring-cloud-sleuth-core/src/main/java/org/springframework/cloud/sleuth/instrument/reactor/ReactorSleuth.java#L76-L79

robotmrv replied:
sorry, do not get it.  
what do you propose?

adriancole commented on spring-cloud-sleuth-core/src/main/java/org/springframework/cloud/sleuth/instrument/web/TraceWebFilter.java:
```diff
@@ -235,18 +264,40 @@ private Span findOrCreateSpan(Context c) {
+
+			@Override
+			public void cancel() {
+				try (Scope scope = this.currentTraceContext.maybeScope(span.context())) {
```
do we expect overhead like this to be worth it? operations that quickly return? I'm not saying I think it shouldn't but each of  "trigger type commands" you can ask the question as if for example, the act of doing that does only small local code it can be unhelpful to wrap.

adriancole commented on spring-cloud-sleuth-core/src/main/java/org/springframework/cloud/sleuth/instrument/web/TraceWebFilter.java:
```diff
@@ -256,14 +307,18 @@ public void onNext(Void aVoid) {
 			public void onError(Throwable t) {
-				terminateSpan(t);
-				this.actual.onError(t);
+				try (Scope scope = this.currentTraceContext.maybeScope(span.context())) {
```
in brave typically, we've done like you are doing here.. decorate the onXXX hooks, but not `request` or `cancel`. 

That said there is not a huge amount of feedback for the impact of not doing  `request` or `cancel` either. Just each of these things has a chance to have a bug, so doing less means less weird unit tests to write and keep passing.

adriancole commented on spring-cloud-sleuth-core/src/main/java/org/springframework/cloud/sleuth/instrument/web/client/TraceWebClientBeanPostProcessor.java:
```diff
@@ -277,6 +280,14 @@ public Context currentContext() {
+			if (key == Scannable.Attr.RUN_STYLE) {
+				return Scannable.Attr.RunStyle.SYNC;
+			}
+			return null;
```
maybe trailing comment why null or add Nullable back to the override for clarity. That way it is easy to know if this is a bug or not even if less familiar

adriancole reviewed (commented):
This is an interesting way to reduce the restacking of the same context and I like that the lower libs have managed this! Thanks for the help addressing a common complaint.

I think there's a large unit test burden less about the code you are adding, mostly general absence before you came. It is probably a good time to add some to cover the use cases mentioned in the comments.

cc @minwoox on impact of this and armeria

marcingrzejszczak commented:
Closing this since we're suggesting the manual reactor instrumentation mode.

robotmrv commented:
@marcingrzejszczak  
manual reactor instrumentation mode could cover only code which is controlled by user, what about framework and 3rd party libs logs?  
`onEach` mode could handle them, why not try to improve its performance?  
I could rebase this PR and add tests if you would like to continue.

marcingrzejszczak commented:
You can wrap the calls to frameworks and logging (or will we provide proper abstractions e.g. in Gateway for better logging). Of course if you're interested you can improve the performance of the ON_EACH approach. Still, for now, we think that the best, suggested approach is to use the MANUAL instrumentation mode.

robotmrv commented:
@marcingrzejszczak  @adriancole  
I've rebased this PR, added tests and addressed review comments in new PR https://github.com/spring-cloud/spring-cloud-sleuth/pull/1830  
please take a look
